dxr/plugins/descriptor/__init__.py

```diff
@@ -0,0 +1,150 @@
+"""Descriptor - Provide useful descriptions for common file types.
+Refer to https://mxr.mozilla.org/webtools-central/source/mxr/Local.pm#27
+"""
+from itertools import ifilter
+import re
+from os import listdir
+from os.path import splitext, basename, join, isfile
+from warnings import warn
+
+import dxr.indexers
+
+
+def is_readme(filename):
+    """Return whether filename is probably a readme."""
+    return filename.lower() in {'readme', 'readme.md', 'readme.txt'}
+
+
+def describe_readme(lines):
+    """Return a string that represents a description for the given lines
+    of a presumed readme file or None if it can extract no suitable
+    description."""
+    # For now the heuristic is just the first non-empty line.
+    return next(ifilter(None, (line.strip() for line in lines)), None)
+
+
+class FolderToIndex(dxr.indexers.FolderToIndex):
+    browse_headers = ['Description']
+
+    def needles(self):
+        """If the folder contains a readme, then yield the first line of the
+        readme as the description.
+
+        Similar to https://mxr.mozilla.org/webtools-central/source/mxr/Local.pm#251.
+        """
+        for entry in sorted(listdir(self.path)):
+            try:
+                entry = entry.decode('utf-8')
+            except UnicodeDecodeError:
+                continue
+            path = join(self.path, entry)
+            # If we find a readme, then open it and return the first line if
+            # it's non-empty.
+            if is_readme(entry) and isfile(path):
+                with open(path) as readme:
+                    try:
+                        first_line = readme.readline(100).decode(self.tree.source_encoding)
+                    except UnicodeDecodeError:
+                        continue
+                    description = describe_readme([first_line])
+                    if description:
+                        # Pack into a list for consistency with the file needle.
+                        return [("Description", [description])]
+        # Didn't find anything to use as a description
+        return []
+
+
+class FileToIndex(dxr.indexers.FileToIndex):
+    """Do lots of work to yield a description needle."""
+
+    comment_re = re.compile(r'(?:.*?/\*+)(?:\s*\*?\s*)(?P<description>.*?)(?:(?:\*+/)|(?:$))', flags=re.S)
+    docstring_res = [re.compile(r'"""\s*(?P<description>[^"]*)', flags=re.M),
```

Comment:
When I said…

> You can just scan until anything but a quote. That'll greatly reduce the max length of the rewind in most cases.

…I didn't mean "and then stop". I meant "and then see if it was the first quote of a """.